api/models/threadReaction.js

```diff
@@ -0,0 +1,134 @@
+// @flow
+import { db } from './db';
+import {
+  sendThreadReactionNotificationQueue,
+  processReputationEventQueue,
+} from 'shared/bull/queues';
+import type { DBThreadReaction } from 'shared/types';
+import { events } from 'shared/analytics';
+import { trackQueue } from 'shared/bull/queues';
+
+type ReactionType = 'like';
+
+// prettier-ignore
+export const getThreadReactions = (threadIds: Array<string>): Promise<Array<DBThreadReaction>> => {
+  const distinctMessageIds = threadIds.filter((x, i, a) => a.indexOf(x) == i);
+  return db
+    .table('threadReactions')
+    .getAll(...distinctMessageIds, { index: 'threadId' })
+    .filter(row => row.hasFields('deletedAt').not())
+    .group('threadId')
+    .run();
+};
+
+type ThreadReactionInput = {
+  threadId: string,
+  type: ReactionType,
+};
+
+// prettier-ignore
+export const addThreadReaction = (input: ThreadReactionInput, userId: string): Promise<DBThreadReaction> => {
+  return db
+    .table('threadReactions')
+    .getAll(input.threadId, { index: 'threadId' })
+    .filter({ userId })
+    .run()
+    .then(async results => {
+      // if the reaction already exists in the db, it was previously deleted
+      // just remove the deletedAt field
+      if (results && results.length > 0) {
+        const thisReaction = results[0];
+
+        trackQueue.add({
+          userId,
+          event: events.THREAD_REACTION_CREATED,
+          context: {
+            threadReactionId: thisReaction.id,
+          },
+        });
+
+        sendThreadReactionNotificationQueue.add({ threadReaction: thisReaction, userId });
```

Comment:
This is probably not ideal? If I'm reading this correctly, it would mean that if I spam-click the thread reaction button you'll be sent millions of notifications, one for each click. We should only ever send one of 'em?

Reply:
This is handled in athena. It works by looking up a previous notification of the same type on the same entity in the previous 30mins, then will de-duplicate actors. So the result is that it will mark the notification as "unread" for the thread author, but will not actually insert multiple notifications unless there is a 30 min gap in between.